src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionTelemetry.cs

```diff
@@ -0,0 +1,214 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.Runtime.InteropServices;
+using System.Threading;
+using Microsoft.Extensions.Internal;
+
+namespace System.Net
+{
+    [EventSource(Name = "System.Net.NameResolution")]
+    internal sealed class NameResolutionTelemetry : EventSource
+    {
+        public static readonly NameResolutionTelemetry Log = new NameResolutionTelemetry();
+
+        public static new bool IsEnabled => Log.IsEnabled();
+
+        private const int ResolutionStartEventId = 1;
+        private const int ResolutionSuccessEventId = 2;
+        private const int ResolutionFailureEventId = 3;
+
+        private PollingCounter? _lookupsRequestedCounter;
+        private EventCounter? _lookupsDuration;
+
+        private long _lookupsRequested;
+
+        protected override void OnEventCommand(EventCommandEventArgs command)
+        {
+            if (command.Command == EventCommand.Enable)
+            {
+                // The cumulative number of name resolution requests started since the process started.
+                _lookupsRequestedCounter ??= new PollingCounter("dns-lookups-requested", this, () => Interlocked.Read(ref _lookupsRequested))
+                {
+                    DisplayName = "DNS Lookups Requested"
+                };
+
+                _lookupsDuration ??= new EventCounter("dns-lookups-duration", this)
+                {
+                    DisplayName = "Average DNS Lookup Duration",
+                    DisplayUnits = "ms"
+                };
+            }
+        }
+
+        private const int MaxIPFormattedLength = 128;
+
+        [Event(ResolutionStartEventId, Level = EventLevel.Informational)]
```

Comment:
There should be a separate void-returning private method that actually does the WriteEvent and that has the right signature for the data being traced.

Reply:
Is that really needed? I took inspiration from [AspNetCore](https://github.com/dotnet/aspnetcore/blob/master/src/SignalR/common/Http.Connections/src/Internal/HttpConnectionsEventSource.cs#L57-L68)

Reply:
This pattern is very non-standard.

Reply:
There are a few things that are odd about this event.  In general, you want your events to be of the form:

```
[Event(id,...)]
public void EventName(args...)
{
    WriteEvent(id, args...);
}
```

If the payload arguments are expensive to compute, you can call `EventSource.IsEnabled` before calling the event method.  If you want to encapsulate this all into one method, you can do so by creating a method and marking it as [NonEvent].

It's especially important that when you do define events that their method signature matches the set of arguments that you pass to `WriteEvent`.  If this invariant isn't maintained then the events won't be consumable by analysis tools.  `EventSource` does work to emit metadata about the events, and that metadata is based on the method signature of each event that has the `[Event]` attribute attached to it.

Reply:
I assumed that if the signature of the `[Event]` method is string, passing a `ROS<char>` manually would be transparent to EventSource.
Are you saying this event should have a signature of `ROS<char>`, and then have either a string/span passed to it? Are Spans recognized types by EventSource?

Reply:
As of today, `Span` types are not handled by `EventSource`, so you would need to convert it to a type that is handled by `EventSource`, such as `string`, an array, etc.  I think I understand what you're trying to do now - you should just have a `[NonEvent]` method that does the convert from `Span` to `string`, and then have it call the `[Event]` method passing the string.

@noahfalk, I think adding support for `Span` types to be converted into supported logging types would be a worthy thing to do.

Reply:
The idea behind all the Span/string logic was precisely to avoid the Span to string conversion.

Is there an issue behind passing a `ROS<char>` in the same format a string would be passed, if the same event ID is used?

Reply:
I see.  I think this will work.  Though, we really should implement this support in the platform so that you don't have to do this.  I'd recommend putting some comments around this so that it's clear what you're doing when someone comes along and finds this in a year or two.

Reply:
Changed. The `[Event]` methods are now void-returning and `[NonEvent]` code calls into them when possible.
The `IPAddress` codepath calls into custom `WriteEvent` overloads with the same ID and data format to avoid an extra string allocation.